### ISSUE TITLE
Allowing for state change after each step

### ISSUE THREAD

**Issue by marcelvangerven** [patrick-kidger/diffrax#277]
In some settings we want to reset the state of the system after each solver step depending on the state. E.g. when performing a voltage reset in a leaky integrate-and-fire model. This seems unsupported at the moment. The most logical route seems to be to add support for an AbstractStateChangingEvent (similar to DiscreteTerminatingEvent) which changes the state but does not terminate. Would it be possible to add this?

The alternative used now is to overload the solver as in the example below but this is very clunky:

class CustomEuler(Euler):
    """
    Custom solver. Needed to implement reset of the membrane potential and recording of spikes for the LIF model   
    """

    Vthresh: float
    EL: float

    def step(
        self,
        terms: AbstractTerm,
        t0: Scalar,
        t1: Scalar,
        y0: PyTree,
        args: PyTree,
        solver_state: _SolverState,
        made_jump: Bool,
    ) -> Tuple[PyTree, _ErrorEstimate, DenseInfo, _SolverState, RESULTS]:
       
        y1, _, dense_info, _, RESULTS.successful = super().step(terms, t0, t1, y0, args, solver_state, made_jump)

        V = y1[0]

        # membrane potential reset
        spikes = jnp.float32(V > self.Vthresh)
        Vnew = (1.0 - spikes) * V + spikes * self.EL
    
        return (Vnew, spikes), None, dense_info, None, RESULTS.successful

**Comment by patrick-kidger** [comment]
Right, this would be most neatly handled by extending our event handling. I've actually got a comment in the source code [here](https://github.com/patrick-kidger/diffrax/blob/2638a3534b846d10225273e615bbf920f934cc3c/diffrax/integrate.py#L270) about where we should do this.

Realistically this is a nontrivial amount of work to stand up, so it's not something I've ever gotten around to.

Right now I'd recommend using your current approach, or -- if you're feeling ambitious -- I'd be happy to shepherd a PR on this.
